### PR TITLE
[ecovacs] Allow loading device descriptions from a user file

### PR DESCRIPTION
**Pull request by maniac103**
If a new device is released within our release cycle, getting the description for that device into the binding was impossible for users before. With this change, they can create an override file in their installation (with contents we'll need to supply to them), which serves as a stop-gap solution until the next OH release.

Additionally, add the definition for the device that triggered this change to the supported device list.

### CODE REVIEW

File: bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/api/impl/EcovacsApiImpl.java [hunk 1]
```diff
@@ -188,29 +192,58 @@ public List<EcovacsDevice> getDevices() throws EcovacsApiException, InterruptedE
...
+        Path customDescsPath = Paths.get(OpenHAB.getUserDataFolder(), "ecovacs").resolve("custom_device_descs.json");
+        if (Files.exists(customDescsPath)) {
+            try (Reader reader = Files.newBufferedReader(customDescsPath)) {
+                int builtins = descs.size();
+                for (DeviceDescription desc : loadSupportedDeviceData(reader)) {
+                    if (descs.containsKey(desc.deviceClass)) {
+                        logger.trace("Overriding built-in description for {} with custom description",
+                                desc.deviceClass);
+                    }
+                    descs.put(desc.deviceClass, desc);
```
Comment (jlaur) [inline bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/api/impl/EcovacsApiImpl.java:217]:
```suggestion
                    DeviceDescription builtinDesc = descs.put(desc.deviceClass, desc);
                    if (builtinDesc != null) {
                        logger.trace("Overriding built-in description for {} with custom description",
                                desc.deviceClass);
                    }
```

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks! Only one minor suggestion.

_EDIT:_ Sorry, forgot to mention that README should be updated also.

**Comment by maniac103** [comment]
> Sorry, forgot to mention that README should be updated also.

Done, please let me know what you think of it. I've intentionally been a bit vague on the full JSON object format, because it's unlikely that a device list entry will be sufficient in case of completely new models.

File: bundles/org.openhab.binding.ecovacs/README.md [hunk 2]
```diff
@@ -173,3 +173,20 @@ Bridge ecovacs:ecovacsapi:ecovacsapi [ email="your.email@provider.com", password
...
+## Adding support for unsupported models
+
+When encountering an unsupported model during discovery, the binding creates a log message like this one:
+
+```
+2023-04-21 12:02:39.607 [INFO ] [acs.internal.api.impl.EcovacsApiImpl] - Found unsupported device DEEBOT N8 PRO CARE (class s1f8g7, company eco-ng), ignoring.
+```
+
+In such a case, please [create an issue on GitHub](https://github.com/openhab/openhab-addons/issues), listing the contents of the log line.
+In addition to that, If the model is similar to an already supported one, you can try to add the support for yourself (until getting an updated binding).
```
Comment (jlaur) [inline bundles/org.openhab.binding.ecovacs/README.md:185]:
```suggestion
In addition to that, if the model is similar to an already supported one, you can try to add the support yourself (until getting an updated binding).
```

File: bundles/org.openhab.binding.ecovacs/README.md [hunk 3]
```diff
@@ -173,3 +173,20 @@ Bridge ecovacs:ecovacsapi:ecovacsapi [ email="your.email@provider.com", password
...
+## Adding support for unsupported models
+
+When encountering an unsupported model during discovery, the binding creates a log message like this one:
+
+```
+2023-04-21 12:02:39.607 [INFO ] [acs.internal.api.impl.EcovacsApiImpl] - Found unsupported device DEEBOT N8 PRO CARE (class s1f8g7, company eco-ng), ignoring.
+```
+
+In such a case, please [create an issue on GitHub](https://github.com/openhab/openhab-addons/issues), listing the contents of the log line.
+In addition to that, If the model is similar to an already supported one, you can try to add the support yourself (until getting an updated binding).
```
Comment (jlaur) [inline bundles/org.openhab.binding.ecovacs/README.md:185]:
```suggestion
In addition to that, if the model is similar to an already supported one, you can try to add the support yourself (until getting an updated binding).
```

File: bundles/org.openhab.binding.ecovacs/README.md [hunk 4]
```diff
@@ -173,3 +173,20 @@ Bridge ecovacs:ecovacsapi:ecovacsapi [ email="your.email@provider.com", password
...
+For doing so, you can follow the following steps:
+
+- in the `<OPENHAB_USERDATA>/evocacs` folder, create a file named `custom_device_descs.json`
```
Comment (jlaur) [inline bundles/org.openhab.binding.ecovacs/README.md:188]: I assume this directory is not created automatically by the binding, so the user must create it?

Comment (maniac103) [reply bundles/org.openhab.binding.ecovacs/README.md]: Yes. I've added it as first step, I assume that was what you had in mind?

**Review by jlaur — APPROVED** [review]
LGTM

**Pull request merged**